backend/layers/persistence/persistence.py

```diff
@@ -507,6 +507,46 @@ def get_all_datasets(self) -> Iterable[DatasetVersion]:
             dataset_version_ids.extend(collection.datasets)
         return self._get_datasets(dataset_version_ids)
 
+    def get_all_published_collections_with_datasets(self) -> List[CollectionVersionWithDatasets]:
```

Comment:
Why do we need another method in persistence? In fact, even the business layer has two methods that should give you all you need: `get_collections` and `get_all_published_datasets`. You can just call those two and join them by id. This will also guarantee that the /datasets index endpoint will return the same exact values than the portal API version.

Reply:
yeah this is a good point--fwiw, even this implementation as written is business logic mapping results from `get_all_mapped_collection_versions` and `_get_datasets` so it should be in the business layer

Reply:
@ebezzi I'll look closer at those methods and see if I can refactor.

@nayib-jose-gloria I disagree based on the precedents set by other methods. There really is no "business" logic inherent to this API call. It is simply grabbing entities, and then there's a little api-layer logic which is restructuring the response object.

Reply:
@danieljhegeman I meant API layer, I edited my response after I sent it

Reply:
Looked at the code again, and remembered why I chose to implement in a single persistence method:

using @ebezzi 's suggestion of `get_collections` and `get_all_published_datasets` would perform the 'collections' lookup twice, because it happens in both of those calls. My method streamlines this and only fetches all collections once.

Reply:
Rather than jumping from a list of collections with datasets to a list of datasets with collections, go straight to a list of datasets with collections from the persistence layer. The name would be `get_all_published_datasets_with_collection`.

Reply:
There has gotta be a query we can do that does most of this filtering on the database side.

Reply:
So I'm trying to follow the precedents set in the code base already, which means not using sql alchemy joins (I'm assuming that's what you're suggesting @Bento007 ). I'll wait for consensus from @ebezzi and @nayib-jose-gloria.

Reply:
One more suggestion would be to modify `get_all_mapped_collection_versions` to optionally return `Iterable[CollectionVersionWithDatasets]` 

```
    def get_all_mapped_collection_versions(self, get_tombstoned: bool = False, datasets: bool =True) -> Union[Iterable[
        CollectionVersion], Iterable[CollectionVersionWithDatasets]]:
        .
        .
        .
            for version in versions:
                canonical_row = mapped_version_ids[version.id]
                canonical = CanonicalCollection(
                    CollectionId(str(canonical_row.id)),
                    CollectionVersionId(str(canonical_row.version_id)),
                    canonical_row.originally_published_at,
                    canonical_row.revised_at,
                    canonical_row.tombstone,
                )
                if datasets:
                    _datasets = self._get_datasets([DatasetVersionId(str(id)) for id in version.datasets])
                    yield self._row_to_collection_version_with_datasets(version, canonical, _datasets)
                else:
                    yield self._row_to_collection_version(version, canonical)

Reply:
Thanks for the suggestion @Bento007 . I definitely want to go with more code reuse. I think I would rather just always return both the collections and datasets. The user can ignore one, use the other, or use both. Slightly more readable/usable that way.

Reply:
you can also go the extra mile and add some caching over getting the collection and dataset_version

Reply:
@danieljhegeman in the spirit of reducing the number of needless calls though, there are use cases leveraging this now that have no need to make the db calls to populate the extra dataset info

Reply:
Good point.

Reply:
>you can also go the extra mile and add some caching over the collection and dataset versions
probably not since one of the general operating principles for the Discover API has been 'most recent data always'. So we'd need to track most-recently-published version in some way. A feature for another day, perhaps.

Reply:
you cache on the version_id and that would all be handled by the LRU cache, but it does not need to be part of this PR.